components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java

```diff
@@ -264,58 +260,107 @@ public SubscriptionResponse addApiSubscription(String apiId, String applicationI
             workflow.setApiVersion(api.getVersion());
             workflow.setApiProvider(api.getProvider());
-
             workflow.setApiId(apiId);
-            
             workflow.setTierName(tier);
-            
-
             workflow.setApplicationName(application.getName());
             workflow.setApplicationId(applicationId);
-            
             workflow.setSubscriber(getUsername());
+
             WorkflowResponse response = addSubscriptionWFExecutor.execute(workflow);
             workflow.setStatus(response.getWorkflowStatus());
-            
-            addWorkflowEntries(workflow);          
 
-            
+            addWorkflowEntries(workflow);
+
             if (WorkflowStatus.APPROVED == response.getWorkflowStatus()) {
                 completeWorkflow(addSubscriptionWFExecutor, workflow);
-            }      
+            }
 
             subScriptionResponse = new SubscriptionResponse(subscriptionId, response);
-            
-            
+
         } catch (APIMgtDAOException e) {
             String errorMsg = "Error occurred while adding api subscription for api - " + apiId;
             log.error(errorMsg, e);
             throw new APIManagementException(errorMsg, e, ExceptionCodes.APIMGT_DAO_EXCEPTION);
         }
- 
+
         return subScriptionResponse;
     }
 
     @Override
     public void deleteAPISubscription(String subscriptionId) throws APIManagementException {
         try {
-            getApiSubscriptionDAO().deleteAPISubscription(subscriptionId);
+
+            WorkflowExecutor createSubscriptionWFExecutor = WorkflowExecutorFactory.getInstance()
+                    .getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_SUBSCRIPTION_CREATION);
+            WorkflowExecutor removeSubscriptionWFExecutor = WorkflowExecutorFactory.getInstance()
+                    .getWorkflowExecutor(WorkflowConstants.WF_TYPE_AM_SUBSCRIPTION_DELETION);
+
+            // check for pending subscription creation
+            if (subscriptionId == null) {
+                String errorMsg = "Subscription Id is not provided";
+                log.error(errorMsg);
+                throw new APIManagementException(errorMsg, ExceptionCodes.PARAMETER_NOT_PROVIDED);
+            }
+            
+            Subscription subscription = getApiSubscriptionDAO().getAPISubscription(subscriptionId);
+            if (subscription == null) {
+                String errorMsg = "Subscription not found for the id - " + subscriptionId;
+                log.error(errorMsg);
+                throw new APIManagementException(errorMsg, ExceptionCodes.SUBSCRIPTION_NOT_FOUND);
+            } else {
+                if (APIMgtConstants.SubscriptionStatus.ON_HOLD == subscription.getStatus()) {
+                    String pendingRefForSubscription = getWorkflowDAO()
+                            .getExternalWorkflowReferenceForSubscription(subscriptionId);
+                    if (pendingRefForSubscription != null) {
```

Comment:
if you take the overall feature, this reference number is generated by the code. It is a random number. so this value is either null (if there is no entry in db) or some random value. There is no chance of this value becoming an empty one.